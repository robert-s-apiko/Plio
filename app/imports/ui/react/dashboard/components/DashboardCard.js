import PropTypes from 'prop-types';
import React from 'react';
import { Card, CardBody, CardTitle, CardFooter, CardImg } from 'reactstrap';
import cx from 'classnames';
import styled from 'styled-components';

import { CardBlock, Icon } from '../../components';

const Wrapper = styled.div`
  min-height: 92px;
`;

const DashboardCard = ({
  className,
  disabled,
  title,
  footer,
  icon,
  text,
  label,
  ...props
}) => (
  <Card
    tag="a"
    className={cx(
      className,
      'dashboard-item',
      { 'dashboard-item-disabled': disabled },
    )}
    {...props}
  >
    <CardBlock>
      <Wrapper>
        <CardImg
          top
          tag={Icon}
          className="dashboard-item-icon"
          name={cx('fw', icon)}
        />
        <CardBody>
          <CardTitle>
            {title}
          </CardTitle>
        </CardBody>
      </Wrapper>
      <CardFooter>
        <span className="text-muted">{text}</span>
        {label && (<span className="label label-danger margin-left">{label}</span>)}
      </CardFooter>
    </CardBlock>
  </Card>
);

DashboardCard.propTypes = {
  className: PropTypes.string,
  disabled: PropTypes.bool,
  icon: PropTypes.string,
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  text: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  title: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  footer: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
};

export default DashboardCard;
