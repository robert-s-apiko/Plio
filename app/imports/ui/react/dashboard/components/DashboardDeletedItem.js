import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { compose, withProps, onlyUpdateForKeys, withHandlers } from 'recompose';
import { Button, ButtonGroup } from 'reactstrap';
import { getFormattedDate } from '../../../../share/helpers';
import Label from '../../components/Labels/Label';

const StyledDeletedGoal = styled.div`
  cursor: inherit;
  .btn-group {
    position: relative;
    top: -2px;
    float: right;
    .btn.btn-secondary:first-child:not(:last-child) {
      padding-right: .75rem;
    }
  }
`;

const enhance = compose(
  onlyUpdateForKeys(['_id', 'isDeleted', 'completedAt', 'deletedAt', 'title', 'sequentialId']),
  withHandlers({
    onRestore: ({
      _id,
      isDeleted,
      onUndoCompletion,
      onRestore,
    }) => () => isDeleted ? onRestore(_id) : onUndoCompletion(_id),
    onRemove: ({ _id, title, onRemove }) => () => onRemove({ _id, title }),
  }),
  withProps(({ canRestore, ...props }) => ({
    date: props.isDeleted ? props.deletedAt : props.completedAt,
    statusText: props.isDeleted ? 'deleted' : 'completed',
    canRestore: canRestore(props),
  })),
);

const DashboardDeletedItem = ({
  sequentialId,
  title,
  date,
  statusText,
  onRestore,
  onRemove,
  canRestore,
}) => (
  <StyledDeletedGoal className="dashboard-stats-action">
    <Label>{sequentialId}</Label>
    {title}
    <span className="text-muted"> ({statusText} {getFormattedDate(date)})</span>
    <ButtonGroup size="sm">
      <Button onClick={onRemove}>Delete</Button>
      <Button disabled={!canRestore} onClick={onRestore}>Restore</Button>
    </ButtonGroup>
  </StyledDeletedGoal>
);

DashboardDeletedItem.propTypes = {
  sequentialId: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  statusText: PropTypes.string.isRequired,
  date: PropTypes.number.isRequired,
  onRestore: PropTypes.func.isRequired,
  onRemove: PropTypes.func.isRequired,
  canRestore: PropTypes.bool.isRequired,
};

export default enhance(DashboardDeletedItem);
