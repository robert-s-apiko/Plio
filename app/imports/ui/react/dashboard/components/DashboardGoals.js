import PropTypes from 'prop-types';
import React from 'react';
import pluralize from 'pluralize';
import { joinIds } from 'plio-util';

import {
  DashboardStatsExpandable,
  IconLoading,
  PlusButton,
  GoalsChart,
} from '../../components';
import { GoalAddModalContainer, GoalEditModalContainer } from '../../goals/containers';

const DashboardGoals = ({
  totalCount,
  goals,
  onLineTap,
  toggle,
  isOpen,
  loading,
  isAddModalOpen,
  toggleAddModal,
  userId,
  organizationId,
  isEditModalOpen,
  toggleEditModal,
}) => (
  <DashboardStatsExpandable
    items={goals}
    total={totalCount}
    itemsPerRow={goals.length}
    renderIcon={loading ? () => <IconLoading /> : undefined}
    render={({ items }) => !!items.length && (
      <GoalsChart
        key={joinIds(items)}
        goals={items}
        onScatterTap={toggleEditModal}
        {...{ onLineTap }}
      />
    )}
    {...{ toggle, isOpen }}
  >
    <PlusButton size="1" onClick={toggleAddModal} />
    {pluralize('goal', totalCount || goals.length, true)}
    <GoalAddModalContainer
      isOpen={isAddModalOpen}
      toggle={toggleAddModal}
      ownerId={userId}
      {...{ organizationId }}
    />
    <GoalEditModalContainer
      isOpen={isEditModalOpen}
      toggle={toggleEditModal}
      {...{ organizationId }}
    />
  </DashboardStatsExpandable>
);

DashboardGoals.propTypes = {
  totalCount: PropTypes.number.isRequired,
  goals: PropTypes.arrayOf(PropTypes.object).isRequired,
  // zoomDomain: ?
  onZoom: PropTypes.func,
  onLineTap: PropTypes.func,
  onScatterTap: PropTypes.func,
  toggle: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  isAddModalOpen: PropTypes.bool.isRequired,
  toggleAddModal: PropTypes.func.isRequired,
  loading: PropTypes.bool,
  userId: PropTypes.string.isRequired,
  organizationId: PropTypes.string.isRequired,
  isEditModalOpen: PropTypes.bool.isRequired,
  toggleEditModal: PropTypes.func.isRequired,
};

export default DashboardGoals;
