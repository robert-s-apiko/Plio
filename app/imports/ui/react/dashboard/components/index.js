export { default as DashboardUserStats } from './DashboardUserStats';
export { default as DashboardStatsMessage } from './DashboardStatsMessage';
export { default as DashboardStatsOverdueItems } from './DashboardStatsOverdueItems';
export { default as DashboardStatsUnreadMessages } from './DashboardStatsUnreadMessages';
export { default as DashboardStatsUserList } from './DashboardStatsUserList';
export { default as DashboardStatsAction } from './DashboardStatsAction';
export { default as DashboardGoals } from './DashboardGoals';
