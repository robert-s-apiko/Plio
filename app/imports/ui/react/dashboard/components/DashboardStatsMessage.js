import PropTypes from 'prop-types';
import React from 'react';

import { Icon } from '../../components';

const DashboardStatsMessage = ({
  url,
  extension,
  fullName,
  text,
  timeString,
}) => (
  <a
    className="dashboard-stats-message"
    href={url}
  >
    {extension ? (
      <Icon name={`file-${extension}-o`} />
    ) : (
      <Icon name="comment" />
    )}
    <strong>{fullName}: </strong>
    {text}
    <span className="text-muted">{timeString} ago</span>
  </a>
);

DashboardStatsMessage.propTypes = {
  url: PropTypes.string.isRequired,
  fullName: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  timeString: PropTypes.string.isRequired,
  extension: PropTypes.string,
};

export default DashboardStatsMessage;
