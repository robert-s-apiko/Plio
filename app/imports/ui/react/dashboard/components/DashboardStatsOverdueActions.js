import PropTypes from 'prop-types';
import React from 'react';
import pluralize from 'pluralize';
import { map } from 'ramda';
import { onlyUpdateForKeys } from 'recompose';

import { DashboardStatsExpandable } from '../../components';
import {
  WorkspaceDefaults,
  WorkspaceDefaultsTypes,
} from '../../../../share/constants';
import { DashboardStatsActionContainer } from '../containers';
import { joinIds } from '../../../../client/util';

const enhance = onlyUpdateForKeys(['count', 'workItems', 'orgSerialNumber', 'itemsPerRow']);

export const DashboardStatsOverdueActions = ({
  count,
  workItems,
  orgSerialNumber,
  itemsPerRow = WorkspaceDefaults[WorkspaceDefaultsTypes.DISPLAY_MESSAGES],
}) => (
  <DashboardStatsExpandable
    items={workItems}
    itemsPerRow={itemsPerRow}
    render={({ items }) => (
      <div key={joinIds(items)}>
        {map(item => (
          <DashboardStatsActionContainer key={item._id} {...{ ...item, orgSerialNumber }} />
        ), workItems)}
      </div>
    )}
  >
    {pluralize('overdue work item', count, true)}
  </DashboardStatsExpandable>
);

DashboardStatsOverdueActions.propTypes = {
  workItems: PropTypes.arrayOf(PropTypes.object).isRequired,
  itemsPerRow: PropTypes.number.isRequired,
  orgSerialNumber: PropTypes.number.isRequired,
  count: PropTypes.number.isRequired,
};

export default enhance(DashboardStatsOverdueActions);
