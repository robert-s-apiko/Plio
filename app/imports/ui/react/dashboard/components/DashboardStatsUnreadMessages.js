import PropTypes from 'prop-types';
import React from 'react';
import pluralize from 'pluralize';

import { DashboardStats } from '../../components';
import DashboardStatsUserList from './DashboardStatsUserList';

export const DashboardUserStats = ({
  messages,
  count,
}) => !!messages.length && (
  <DashboardStats>
    <DashboardStats.Title>
      {count}
      
    </DashboardStats.Title>
  </DashboardStats>
);

DashboardUserStats.propTypes = {

};

export default DashboardUserStats;
