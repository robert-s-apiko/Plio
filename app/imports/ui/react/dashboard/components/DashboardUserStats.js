import PropTypes from 'prop-types';
import React from 'react';
import pluralize from 'pluralize';
import { branch } from 'recompose';
import { converge, gt, prop, identity } from 'ramda';

import { getUsersLength } from '../../../../client/util';
import { DashboardStats, Collapse, Button, Icon } from '../../components';
import DashboardStatsUserList from './DashboardStatsUserList';
import { withStateToggle, namedCompose } from '../../helpers';

const usersExceedLimit = converge(gt, [
  getUsersLength,
  prop('usersPerRow'),
]);

const enhance = namedCompose('DashboardUserStats')(
  branch(
    usersExceedLimit,
    withStateToggle(false, 'isOpen', 'toggle'),
    identity,
  ),
);

const DashboardUserStats = enhance(({
  users,
  isOpen,
  toggle,
  usersPerRow,
}) => !!users.length && (
  <DashboardStats>
    <DashboardStats.Title>
      {!!toggle && (
        <Button size="1" color="secondary add" onClick={toggle}>
          <Icon name="plus" />
        </Button>
      )}
      {pluralize('user', users.length, true)} online
    </DashboardStats.Title>
    <DashboardStatsUserList users={users.slice(0, usersPerRow)} />

    {!!toggle && (
      <Collapse {...{ isOpen }}>
        <DashboardStatsUserList users={users.slice(usersPerRow)} />
      </Collapse>
    )}
  </DashboardStats>
));

DashboardUserStats.propTypes = {
  users: PropTypes.arrayOf(PropTypes.object).isRequired,
  usersPerRow: PropTypes.number.isRequired,
};

export default DashboardUserStats;
