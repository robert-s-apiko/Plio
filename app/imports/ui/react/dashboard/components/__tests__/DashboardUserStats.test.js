import React from 'react';
import { mount } from 'enzyme';
import { times } from 'ramda';

import { DashboardUserStats } from '../DashboardUserStats';
import DashboardStatsUserList from '../DashboardStatsUserList';
import { Collapse } from '../../../components';

describe('DashboardUserStats', () => {
  const genUser = n => ({
    _id: n,
    status: 'online',
    profile: {
      firstName: `John ${n}`,
      lastName: `Doe ${n}`,
    },
  });
  it('renders collapse when more than 5 users online', () => {
    const len = 25;
    const users = times(genUser, len);
    const usersPerRow = 5;
    const wrapper = mount(<DashboardUserStats {...{ users, usersPerRow }} />);
    expect(wrapper.find(Collapse)).toHaveLength(1);
    expect(wrapper.find(DashboardStatsUserList)).toHaveLength(len / usersPerRow);
  });

  it('does not render collapse when less than 5 users online', () => {
    const users = times(genUser, 4);
    const usersPerRow = 5;
    const wrapper = mount(<DashboardUserStats {...{ users, usersPerRow }} />);
    expect(wrapper.find(Collapse)).toHaveLength(0);
    expect(wrapper.find(DashboardStatsUserList)).toHaveLength(1);
  });

  it('matches snapshot', () => {
    const users = times(genUser, 25);
    const usersPerRow = 5;
    const wrapper = mount(<DashboardUserStats {...{ users, usersPerRow }} />);
    expect(wrapper).toMatchSnapshot();
  });
});
