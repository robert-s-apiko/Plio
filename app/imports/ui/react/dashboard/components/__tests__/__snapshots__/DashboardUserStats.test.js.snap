// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardUserStats matches snapshot 1`] = `
<DashboardStats>
  <DashboardStatsTitle>
    <defaultProps(mapProps(componentFromProp(component)))
      color="secondary add"
      component="a"
      onClick={[Function]}
      size="1"
    >
      <Icon
        name="plus"
      />
    </defaultProps(mapProps(componentFromProp(component)))>
    25 users
     online
  </DashboardStatsTitle>
  <DashboardStatsUserList
    users={
      Array [
        Object {
          "_id": 0,
          "profile": Object {
            "firstName": "John 0",
            "lastName": "Doe 0",
          },
          "status": "online",
        },
        Object {
          "_id": 1,
          "profile": Object {
            "firstName": "John 1",
            "lastName": "Doe 1",
          },
          "status": "online",
        },
        Object {
          "_id": 2,
          "profile": Object {
            "firstName": "John 2",
            "lastName": "Doe 2",
          },
          "status": "online",
        },
        Object {
          "_id": 3,
          "profile": Object {
            "firstName": "John 3",
            "lastName": "Doe 3",
          },
          "status": "online",
        },
        Object {
          "_id": 4,
          "profile": Object {
            "firstName": "John 4",
            "lastName": "Doe 4",
          },
          "status": "online",
        },
      ]
    }
  />
  <Collapse
    isOpen={false}
  >
    <div
      key=" 5 6 7 8 9"
    >
      <DashboardStatsUserList
        users={
          Array [
            Object {
              "_id": 5,
              "profile": Object {
                "firstName": "John 5",
                "lastName": "Doe 5",
              },
              "status": "online",
            },
            Object {
              "_id": 6,
              "profile": Object {
                "firstName": "John 6",
                "lastName": "Doe 6",
              },
              "status": "online",
            },
            Object {
              "_id": 7,
              "profile": Object {
                "firstName": "John 7",
                "lastName": "Doe 7",
              },
              "status": "online",
            },
            Object {
              "_id": 8,
              "profile": Object {
                "firstName": "John 8",
                "lastName": "Doe 8",
              },
              "status": "online",
            },
            Object {
              "_id": 9,
              "profile": Object {
                "firstName": "John 9",
                "lastName": "Doe 9",
              },
              "status": "online",
            },
          ]
        }
      />
    </div>
    <div
      key=" 10 11 12 13 14"
    >
      <DashboardStatsUserList
        users={
          Array [
            Object {
              "_id": 10,
              "profile": Object {
                "firstName": "John 10",
                "lastName": "Doe 10",
              },
              "status": "online",
            },
            Object {
              "_id": 11,
              "profile": Object {
                "firstName": "John 11",
                "lastName": "Doe 11",
              },
              "status": "online",
            },
            Object {
              "_id": 12,
              "profile": Object {
                "firstName": "John 12",
                "lastName": "Doe 12",
              },
              "status": "online",
            },
            Object {
              "_id": 13,
              "profile": Object {
                "firstName": "John 13",
                "lastName": "Doe 13",
              },
              "status": "online",
            },
            Object {
              "_id": 14,
              "profile": Object {
                "firstName": "John 14",
                "lastName": "Doe 14",
              },
              "status": "online",
            },
          ]
        }
      />
    </div>
    <div
      key=" 15 16 17 18 19"
    >
      <DashboardStatsUserList
        users={
          Array [
            Object {
              "_id": 15,
              "profile": Object {
                "firstName": "John 15",
                "lastName": "Doe 15",
              },
              "status": "online",
            },
            Object {
              "_id": 16,
              "profile": Object {
                "firstName": "John 16",
                "lastName": "Doe 16",
              },
              "status": "online",
            },
            Object {
              "_id": 17,
              "profile": Object {
                "firstName": "John 17",
                "lastName": "Doe 17",
              },
              "status": "online",
            },
            Object {
              "_id": 18,
              "profile": Object {
                "firstName": "John 18",
                "lastName": "Doe 18",
              },
              "status": "online",
            },
            Object {
              "_id": 19,
              "profile": Object {
                "firstName": "John 19",
                "lastName": "Doe 19",
              },
              "status": "online",
            },
          ]
        }
      />
    </div>
    <div
      key=" 20 21 22 23 24"
    >
      <DashboardStatsUserList
        users={
          Array [
            Object {
              "_id": 20,
              "profile": Object {
                "firstName": "John 20",
                "lastName": "Doe 20",
              },
              "status": "online",
            },
            Object {
              "_id": 21,
              "profile": Object {
                "firstName": "John 21",
                "lastName": "Doe 21",
              },
              "status": "online",
            },
            Object {
              "_id": 22,
              "profile": Object {
                "firstName": "John 22",
                "lastName": "Doe 22",
              },
              "status": "online",
            },
            Object {
              "_id": 23,
              "profile": Object {
                "firstName": "John 23",
                "lastName": "Doe 23",
              },
              "status": "online",
            },
            Object {
              "_id": 24,
              "profile": Object {
                "firstName": "John 24",
                "lastName": "Doe 24",
              },
              "status": "online",
            },
          ]
        }
      />
    </div>
  </Collapse>
</DashboardStats>
`;
